{"cells":[{"attachments":{},"cell_type":"markdown","metadata":{},"source":["# Syntrophic interactions in a simplified anaerobic CO2-fixating microbial community\n","## Preliminaries"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["import numpy as np\n","import pandas as pd\n","import cobra as cb\n","import micom as mc\n","import itertools\n","import matplotlib as mpl\n","import matplotlib.pyplot as plt\n","import seaborn as sns\n","\n","import coco as co\n","from constraints import *\n","config = cb.Configuration()\n","config.solver = 'cplex'"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["Set up paths and variables"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["sample_info_path = '../data/sample_info.tsv'\n","dna_path = '../data/all_MAGs_coverage.txt'\n","rna_path = '../data/RNA/'\n","gene_path = '../data/gene_length.csv'\n","model_path = '../models/gapseq/'\n","community_path = '../models/micom/'\n","media_path = '../data/media_db.tsv'\n","community_growth_path = '../data/growth_rates.csv'\n","biochemistry_path = '../data/feedstock.csv'\n","results_path = 'results/'\n","figures_path = 'figures/'\n","build_communities = True\n","abundance_threshold = 0.005"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["## Data preparation"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["Load MAG abundances"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["dna_df = pd.read_table(dna_path, sep='\\t', index_col='Bin Id')\n","mag_size_s = dna_df['Bin size (Mbp)']\n","dna_df = dna_df.iloc[:, dna_df.columns.str.contains('% community', regex=False)]\n","dna_df.columns = dna_df.columns.str.replace('.sorted: % community', '', regex=False)\n","dna_df.index = dna_df.index.str.replace('.fasta', '', regex=False)\n","dna_df = dna_df.loc[dna_df.index != 'unbinned', :]\n","MAG_names = dna_df.index"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["Load sample information and set column names"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["sample_df = pd.read_table(sample_info_path, index_col='NGS ID')\n","dna_df.columns = [s[0] for s in dna_df.columns.str.split('S')]\n","dna_df.columns = sample_df.loc[dna_df.columns, 'ID']"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["Load RNA abundances"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["df_list = []\n","key_list = []\n","for m in MAG_names:\n","    try:\n","        counts = pd.read_table(rna_path + m + '.csv_log2.csv', sep=',', index_col='Unnamed: 0')\n","        counts.index = counts.index.str.replace('.faa_', '__', regex=False)\n","        counts.index = counts.index.str.replace('pilon_pilon', '', regex=False)\n","        counts.index = counts.index.str.replace('_pilon', '', regex=False)\n","        df_list.append(counts)\n","        key_list.append(m)\n","    except:\n","        pass\n","rna_df = pd.concat(df_list, axis=0, keys=key_list)\n","# Set column names\n","rna_df.columns = rna_df.columns.str.replace('1_[ABC][123]', '\\g<0>-s2', regex=True)\n","rna_df.columns = rna_df.columns.str.replace('2_[ABC][123]', '\\g<0>-s3', regex=True)\n","rna_df.columns = rna_df.columns.str.replace('Read_Count_[12]_', '', regex=True)"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["Aggregate and normalise RNA abundances"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# Sum RNA abundances and log-transform\n","tot_rna_df = pd.DataFrame(index=MAG_names, columns=rna_df.columns)\n","for s in rna_df.columns:\n","    count_sums = rna_df[s].groupby(level=0).sum()\n","    tot_rna_df.loc[:, s] = np.log(count_sums)\n","# Load gene lengths\n","gene_df = pd.read_table(gene_path, sep=',', index_col=None)\n","gene_df['gene'] = gene_df['gene'].str.replace('.faa_', '__', regex=False)\n","gene_df['gene'] = gene_df['gene'].str.replace('pilon_pilon', '', regex=False)\n","gene_df['gene'] = gene_df['gene'].str.replace('_pilon', '', regex=False)\n","#gene_df = gene_df.loc[gene_df['gene'].isin(rna_df.index.get_level_values(1)), :]\n","gene_df['MAG'] = gene_df['gene'].str.split('__', expand=True)[0]\n","gene_df.set_index(['MAG', 'gene'], inplace=True)\n","# Sum gene lengths\n","tot_gene_df = gene_df['gene_length'].groupby(level=0).sum().div(1e6)\n","tot_gene_df = tot_gene_df.loc[tot_gene_df.index.isin(rna_df.index.get_level_values(0))]\n","# Normalise RNA abundances with total coding sequence length\n","tot_rna_df = tot_rna_df.divide(tot_rna_df.sum(axis=0))\n","tot_rna_df = tot_rna_df.divide(tot_gene_df, axis=0)\n","tot_rna_df = 100*tot_rna_df.divide(tot_rna_df.sum(axis=0))"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["Select the taxa above the abundance threshold"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# Select samples with both DNA and RNA\n","sample_names = tot_rna_df.columns\n","dna_df = dna_df.loc[:, sample_names]\n","# Combine nucleic acid abundances\n","na_df = dna_df.multiply(tot_rna_df)\n","na_df = na_df.divide(na_df.sum(axis=0))\n","# Filter \n","na_df = na_df.loc[na_df.index != 'unbinned', :]\n","#na_df = na_df.loc[na_df.max(axis=1) >= abundance_threshold, :]\n","MAG_idx = na_df.max(axis=1) >= abundance_threshold\n","MAG_names = na_df.loc[MAG_idx, :].index\n","# Subset RNA tables\n","rna_df = rna_df.loc[MAG_names, :]\n","print('Selected', str(len(MAG_names)), 'MAGs:')\n","print(list(MAG_names))"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["for s in sample_names:\n","    print(s, end=' ')\n","\n","    # Filter species\n","    species_names = list(na_df.index[na_df[s] >= 0.005])\n","    community_abundance = (dna_df.loc[species_names, s]).sum()\n","    print(community_abundance)"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["Adjust MAG names for compatibility with Micom"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["dna_df.index = dna_df.index.str.replace('.', '_', regex=False)\n","rna_df.index = pd.MultiIndex.from_arrays([rna_df.index.get_level_values(0).str.replace('.', '_', regex=False), \\\n","    rna_df.index.get_level_values(1).str.replace('.', '_', regex=False)], names=('MAG', 'gene'))\n","na_df.index = na_df.index.str.replace('.', '_', regex=False)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["sample_names"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["Prepare the media"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["media_df = pd.read_table(media_path, sep='\\t', index_col=None)\n","media_df['excRxn'] = 'EX_' + media_df['met'] + '_m'\n","# dictionary establishing the medium in each sample\n","media_dict = dict(zip(sample_names, np.repeat(['BA[acetate+CO2]'], len(sample_names))))"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["Prepare the biochemical data"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["biochemistry_df = pd.read_table(biochemistry_path, sep=',', index_col=None)\n","biochemistry_df = biochemistry_df.drop('metName', axis=1)\n","biochemistry_df['excRxn'] = 'EX_' + biochemistry_df['met'] + '_m'"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["## Build microbial community models"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["Create community models for each sample"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["if build_communities:\n","    for s in sample_names:\n","        print(s)\n","\n","        # Filter species\n","        species_names = list(na_df.index[na_df[s] >= abundance_threshold])\n","        abundances = list(dna_df.loc[species_names, s])\n","        model_paths = [a+b for a,b in zip([model_path]*len(species_names), species_names)]\n","        model_paths = [a+b for a,b in zip(model_paths, [\".xml\"]*len(species_names))]\n","        model_paths = pd.Series(model_paths).str.replace('metabat1_', 'metabat1.')\n","        model_paths = model_paths.str.replace('metabat2_', 'metabat2.')\n","        model_paths = model_paths.str.replace('maxbin2_', 'maxbin2.')\n","        model_paths = model_paths.str.replace('vamb_', 'vamb.')\n","        model_paths = model_paths.str.replace('metabat1.31', 'metabat1.31-curated')\n","        model_paths = list(model_paths.str.replace('metabat1.59', 'metabat1.59-curated'))\n","        print(species_names)\n","\n","        # Build community model\n","        species_df = pd.DataFrame(data={'id': species_names, 'file': model_paths, 'abundance': abundances})\n","        com = mc.Community(species_df, solver='cplex')\n","        print(\"Built a community with a total of {} reactions.\".format(len(com.reactions)))\n","        print(\"Built a community with a total of {} metabolites.\".format(len(com.metabolites)))\n","\n","        # Save community model\n","        com.to_pickle(community_path + s + '_' + str(abundance_threshold) + '.pickle')"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["## Parameter exploration\n","Explore parameters for CoCo modelling. "]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["First, initialise CoCo"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["cocoGEM_builder = co.CoCo(rna_df, sparse_community=True)"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["Parameter exploration"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["deltas = [0.5, 0.6, 0.7, 0.8, 0.9, 1., 2., 3., 4., 5.] #np.arange(1., 11., 1.)\n","gammas = [1] #np.arange(1., 3)\n","tradeoffs = np.round(np.arange(0.1, 1.05, 0.1), 1) # cooperative tradeoff fraction\n","label = ''\n","method='ct'\n","verbose=False\n","\n","param_product_bound, delta_bound = cocoGEM_builder.get_param_range()\n","combinations = np.array([c for c in itertools.product(deltas, gammas, tradeoffs)])\n","param_products = np.array([i[0]*i[1] for i in combinations])\n","combinations = list(map(tuple, combinations[param_products <= param_product_bound]))\n","community_dict = {} # every element is a list of species for a condition\n","exp_growth = []\n","abundances = []\n","base_df_list = []\n","coco_df_list = []\n","base_community_growth = pd.DataFrame(index=sample_names, columns=pd.MultiIndex.from_tuples(combinations))\n","coco_community_growth = pd.DataFrame(index=sample_names, columns=pd.MultiIndex.from_tuples(combinations))\n","\n","print('deltas =', deltas)\n","print('gammas =', gammas)\n","print('tradeoffs =', tradeoffs)\n","print('method =', method)\n","\n","\n","for s in sample_names:\n","    print('----------------------------------')\n","    print('>\\t', s)\n","    print('----------------------------------')\n","\n","    # Filter species\n","    species_names = list(na_df.index[na_df[s] >= abundance_threshold])\n","    abundances.append(list(dna_df.loc[species_names, s]))\n","    community_dict[s] = species_names\n","    community_abundance = (na_df.loc[na_df[s] >= abundance_threshold, s]).sum()\n","    print(species_names)\n","\n","    # Load community model\n","    com = mc.load_pickle(community_path + s + '_' + str(abundance_threshold) + '.pickle')\n","    print(\"Built a community with a total of {} reactions.\".format(len(com.reactions)))\n","    print(\"Built a community with a total of {} metabolites.\".format(len(com.metabolites)))\n","\n","    # fixes\n","    if 'metabat1_59' in species_names:\n","        com.reactions.rxn90070_c0__metabat1_59.bounds = 0.0, 1000.0\n","        com.reactions.rxn90072_c0__metabat1_59.bounds = -1000.0, 0.0\n","    \n","    # apply environmental constraints\n","    com = set_medium(com, s, media_dict, media_df)\n","    com = set_biochemical_constraints(com, s, biochemistry_df, community_abundance)\n","    \n","    # gene coverage check\n","    _ = cocoGEM_builder.check_gene_coverage(com, verbose=True)\n","\n","    # simulations over delta, gamma and tradeoff\n","    base_growths = pd.DataFrame(index=species_names, columns=pd.MultiIndex.from_tuples(combinations))\n","    coco_growths = pd.DataFrame(index=species_names, columns=pd.MultiIndex.from_tuples(combinations))\n","    for t in tradeoffs:\n","        try:\n","            if method == 'ct':\n","                sol = com.cooperative_tradeoff(fraction=t, fluxes=False, min_growth=0.01)\n","            elif method == 'fba':\n","                sol = com.optimize_all()\n","            for c in combinations:\n","                if c[2] == t:\n","                    if method == 'ct':\n","                        base_growths.loc[:, c] = list(sol.members.loc[species_names, 'growth_rate'])\n","                        base_community_growth.loc[s, c] = sol.growth_rate\n","                    elif method == 'fba':\n","                        base_growths.loc[:, c] = list(sol.loc[species_names])\n","        except:\n","            print('No solution')\n","    print('-----------------')\n","\n","    for c in combinations:\n","        with com as coco_com:\n","            print('delta = ' + str(c[0]) + ', gamma = ' + str(c[1]) + ', tradeoff = ' + str(c[2]))\n","            cocoGEM_builder.build(coco_com, s, c[0], c[1], verbose=verbose)\n","            print('-----------------')\n","            try:\n","                if method == 'ct':\n","                    sol = coco_com.cooperative_tradeoff(fraction = c[2], fluxes=False)\n","                    coco_growths.loc[:, c] = list(sol.members.loc[species_names, 'growth_rate'])\n","                    coco_community_growth.loc[s, c] = sol.growth_rate\n","                elif method == 'fba':\n","                    sol = coco_com.optimize_all()\n","                    coco_growths.loc[:, c] = list(sol.loc[species_names])\n","            except:\n","                print('No solution')\n","    base_df_list.append(base_growths)\n","    coco_df_list.append(coco_growths)\n","\n","    print('\\n')\n","print('Finished')\n","\n","# wrap up results\n","base_growth = pd.concat(base_df_list, axis=0, keys=sample_names)\n","coco_growth = pd.concat(coco_df_list, axis=0, keys=sample_names)\n","abundances = pd.Series(np.hstack(abundances))\n","\n","# save\n","base_growth.to_csv(results_path + 'base_growths_' + label + '.csv')\n","coco_growth.to_csv(results_path + 'coco_growths_' + label + '.csv')\n","base_community_growth.to_csv(results_path + 'base_community_growths_' + label + '.csv')\n","coco_community_growth.to_csv(results_path + 'coco_community_growths_' + label + '.csv')\n","abundances.to_csv(results_path + 'abundances.csv')"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["Extract the parameters yielding the best correlations with community growth"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["exp_community_growth = pd.read_table(community_growth_path, sep=',', index_col='sample')\n","exp_community_growth = exp_community_growth.T.squeeze()\n","exp_community_growth"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["from scipy.stats import pearsonr\n","\n","label = ''\n","\n","base_growth_df = base_community_growth.drop(['A2-s3'], axis=0)\n","coco_growth_df = coco_community_growth.drop(['A2-s3'], axis=0)\n","exp_growth_s = exp_community_growth.drop(['A2-s3'], axis=0)\n","\n","base_growth_df[base_growth_df.abs() < 1e-6] = 0\n","coco_growth_df[coco_growth_df.abs() < 1e-6] = 0\n","\n","# determine the best correlations\n","base_corrs = pd.DataFrame(index=['pearsonr','pvalue'], columns=base_growth_df.columns)\n","coco_corrs = pd.DataFrame(index=['pearsonr','pvalue'], columns=coco_growth_df.columns)\n","idx = np.array(np.isnan(exp_growth_s))\n","for c in base_growth_df.columns:\n","    if sum(base_growth_df.loc[:, c].isna()) == 0:\n","        base_corrs.loc[:, c] = pearsonr(exp_growth_s[~idx], base_growth_df.loc[~idx, c])\n","for c in coco_growth_df.columns:\n","    if sum(coco_growth_df.loc[:, c].isna()) == 0:\n","        coco_corrs.loc[:, c] = pearsonr(exp_growth_s[~idx], coco_growth_df.loc[~idx, c])\n","best_base_idx = pd.to_numeric(base_corrs.loc['pearsonr']).idxmax()\n","best_coco_idx = pd.to_numeric(coco_corrs.loc['pearsonr']).idxmax()\n","best_corrs = [base_corrs.loc['pearsonr', best_base_idx], \\\n","    coco_corrs.loc['pearsonr', best_coco_idx]]\n","best_pvals = [base_corrs.loc['pvalue', best_base_idx], \\\n","    coco_corrs.loc['pvalue', best_coco_idx]]\n","\n","# format data\n","plot_data = pd.DataFrame(data={'Replication rate [a.u.]': list(exp_growth_s) + list(exp_growth_s), \\\n","    'Growth rate [1/h]': np.hstack((base_growth_df.loc[:, best_base_idx], coco_growth_df.loc[:, best_coco_idx])),\n","    'Models': np.array(['Co-GEM']*len(exp_growth_s) + ['CoCo-GEM']*len(exp_growth_s)),\n","    'Conditions': np.tile(np.array(['A']*3 + ['B']*3 + ['C']*3 + ['A']*2 + ['B']*3 + ['C']*3), 2)})\n","plot_data['Growth rate [1/h]'] = plot_data['Growth rate [1/h]'].astype('float')\n","\n","sns.set_style('ticks')\n","mpl.rcParams['axes.labelsize'] = 8\n","mpl.rcParams['xtick.labelsize'] = 8\n","mpl.rcParams['ytick.labelsize'] = 8\n","mpl.rcParams['axes.linewidth'] = 0.8\n","mpl.rcParams['font.sans-serif'] = 'Arial'\n","\n","p = sns.lmplot(data=plot_data.loc[plot_data['Models']=='CoCo-GEM', :], x='Replication rate [a.u.]', y='Growth rate [1/h]', \\\n","    height=1.8, aspect=1.2, scatter_kws={'edgecolors': 'none', 'facecolor': 'slategray', 'linewidths': 0}, line_kws={'color': 'slategray', 'linewidth': 0.8})\n","p.axes[0][0].set_xlim([0, 0.005])\n","p.axes[0][0].set_ylim([0.03, 0.1])\n","p.axes[0][0].set_xlabel('Measured growth rate [1/h]')\n","p.axes[0][0].set_ylabel('Model growth rate [1/h]')\n","p.axes[0][0].tick_params(axis='both', length=2)\n","\n","# set scientific notation on the y axes\n","p.axes[0][0].ticklabel_format(style='scientific', scilimits=(0,0), axis='x')\n","\n","text_kwargs = dict(ha='center', va='center', fontsize=6)\n","plt.text(0.002, 0.09, f\"r = {str(np.around(best_corrs[1], 2))} (p = {str(np.around(best_pvals[1], 4))})\", **text_kwargs)\n","\n","p.fig.savefig(figures_path + 'growth_correlation.pdf', bbox_inches='tight')"]},{"cell_type":"markdown","metadata":{},"source":["## Flux calculation"]},{"cell_type":"markdown","metadata":{},"source":["In case CoCo is not already initialised"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["cocoGEM_builder = co.CoCo(rna_df, sparse_community=True)"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["Calculate the full flux solutions for the best parameter combination"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["delta = 0.9\n","gamma = 1.\n","base_tradeoff = 0.2\n","coco_tradeoff = 0.9\n","label = ''\n","method = 'fba' # 'fba', 'pfba'\n","considered_samples = list(sample_names[0:9])\n","verbose = False\n","\n","community_dict = {} # every element is a list of species for a condition\n","base_flux_df_list = []\n","coco_flux_df_list = []\n","base_exc_df_list = []\n","coco_exc_df_list = []\n","\n","print('delta =', delta)\n","print('gamma =', gamma)\n","print('base_tradeoff =', base_tradeoff)\n","print('coco_tradeoff =', coco_tradeoff)\n","print('method =', method)\n","print('considered_samples =', considered_samples)\n","\n","for s in considered_samples:\n","    print('----------------------------------')\n","    print('>\\t', s)\n","    print('----------------------------------')\n","\n","    # Load community model\n","    com = mc.load_pickle(community_path + s + '_' + str(abundance_threshold) + '.pickle')\n","    species_names = list(com.taxonomy['id'])\n","    community_dict[s] = species_names\n","    rxns = [r.id for r in com.reactions]\n","    excRxns = [e.id for e in com.exchanges]\n","    community_abundance = (na_df.loc[na_df[s] >= abundance_threshold, s]).sum()\n","    print(species_names)\n","    print(\"Built a community with a total of {} reactions.\".format(len(com.reactions)))\n","    print(\"Built a community with a total of {} metabolites.\".format(len(com.metabolites)))\n","\n","    # fixes\n","    if 'metabat1_59' in species_names:\n","        com.reactions.rxn90070_c0__metabat1_59.bounds = 0.0, 1000.0\n","        com.reactions.rxn90072_c0__metabat1_59.bounds = -1000.0, 0.0\n","        com.reactions.rxn90070_c0__metabat1_59.gene_reaction_rule = \\\n","            '( metabat1_59__contig889_2 or metabat1_59__contig889_1239 or metabat1_59__contig889_617 ) and ( metabat1_59__contig889_3 or metabat1_59__contig889_624 )'\n","        com.reactions.rxn90072_c0__metabat1_59.gene_reaction_rule = \\\n","            '( metabat1_59__contig889_2 or metabat1_59__contig889_1239 or metabat1_59__contig889_617 ) and ( metabat1_59__contig889_3 or metabat1_59__contig889_624 )'\n","    if 'metabat1_31' in species_names:\n","        com.reactions.rxn90070_c0__metabat1_31.gene_reaction_rule = \\\n","            '( metabat1_31__contig81_166 or metabat1_31__contig81_167 )'\n","        com.reactions.rxn90072_c0__metabat1_31.gene_reaction_rule = \\\n","            '( metabat1_31__contig81_166 or metabat1_31__contig81_167 )'\n","    \n","    # apply environmental constraints\n","    com = set_medium(com, s, media_dict, media_df)\n","    com = set_biochemical_constraints(com, s, biochemistry_df, community_abundance)\n","    \n","    # gene coverage check\n","    _ = cocoGEM_builder.check_gene_coverage(com, verbose=True)\n","\n","    # simulations\n","    print('--- Base ---')\n","    with com:\n","        try:\n","            sol = com.cooperative_tradeoff(fraction = base_tradeoff, fluxes=False)\n","            med = mc.media.minimal_medium(com, 0.98*sol.growth_rate, min_growth=0.98*sol.members.growth_rate.drop(\"medium\"))\n","            com.medium = med\n","        except:\n","            print('No growth solution')\n","        try:\n","            print('Calculating...')\n","            sol = com.cooperative_tradeoff(fraction = base_tradeoff, fluxes=True, pfba=(method == 'pfba'))\n","        except:\n","            sol = None\n","            print('No solution')\n","        if sol is not None:\n","            print(str(sol.growth_rate))\n","            base_flux_df_list.append(sol.fluxes.loc[species_names, :])\n","            base_exc_df_list.append(sol.fluxes.loc['medium', excRxns])\n","        else:\n","            base_flux_df_list.append(pd.DataFrame(index=species_names, columns=rxns))\n","            base_exc_df_list.append(pd.Series(index=excRxns))\n","\n","    print('--- CoCo ---')\n","    with com as coco_com:\n","        cocoGEM_builder.build(coco_com, s, delta, gamma, meta=True)\n","        \n","        try:\n","            coco_sol = coco_com.cooperative_tradeoff(fraction = coco_tradeoff, fluxes=False)\n","            med = mc.media.minimal_medium(coco_com, 0.98*coco_sol.growth_rate, min_growth=0.98*coco_sol.members.growth_rate.drop(\"medium\"))\n","            coco_com.medium = med\n","        except:\n","            print('No growth solution')\n","        try:\n","            print('Calculating...')\n","            coco_sol = coco_com.cooperative_tradeoff(fraction = coco_tradeoff, fluxes=True, pfba=(method == 'pfba'))\n","        except:\n","            coco_sol = None\n","            print('No solution')\n","        if sol is not None:\n","            print(str(sol.growth_rate))\n","            coco_flux_df_list.append(coco_sol.fluxes.loc[species_names, :])\n","            coco_exc_df_list.append(coco_sol.fluxes.loc['medium', excRxns])\n","        else:\n","            coco_flux_df_list.append(pd.DataFrame(index=species_names, columns=rxns))\n","            coco_exc_df_list.append(pd.Series(index=excRxns))\n","print('Finished')\n","\n","# wrap up\n","base_fluxes = pd.concat(base_flux_df_list, axis=0, keys=considered_samples)\n","coco_fluxes = pd.concat(coco_flux_df_list, axis=0, keys=considered_samples)\n","base_exchanges = pd.DataFrame(base_exc_df_list, index=considered_samples).T\n","coco_exchanges = pd.DataFrame(coco_exc_df_list, index=considered_samples).T\n","\n","# save\n","base_fluxes.to_csv(results_path + 'base_fluxes_' + label + '.csv')\n","coco_fluxes.to_csv(results_path + 'coco_fluxes_' + label + '.csv')\n","base_exchanges.to_csv(results_path + 'base_exchanges_' + label + '.csv')\n","coco_exchanges.to_csv(results_path + 'coco_exchanges_' + label + '.csv')\n"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["Clean values"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["base_fluxes[base_fluxes.abs() < 1e-6] = 0\n","coco_fluxes[coco_fluxes.abs() < 1e-6] = 0\n","base_exchanges[base_exchanges.abs() < 1e-6] = 0\n","coco_exchanges[coco_exchanges.abs() < 1e-6] = 0"]},{"cell_type":"markdown","metadata":{},"source":["Define useful variables"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["considered_samples = sample_names[0:9]\n","\n","# multiply fluxes by the corresponding MAG abundance\n","abundances = []\n","for s in considered_samples:\n","    species_names = list(na_df.index[na_df[s] >= abundance_threshold])\n","    abundances.append(list(dna_df.loc[species_names, s]))\n","abundances = pd.Series(np.hstack(abundances))\n","coco_cum_fluxes = coco_fluxes.loc[considered_samples, :].mul(np.transpose(np.tile(list(abundances/100), (coco_fluxes.shape[1], 1))))\n","\n","A_samples = ['A1-s2', 'A2-s2', 'A3-s2']\n","B_samples = ['B1-s2', 'B2-s2', 'B3-s2']\n","C_samples = ['C1-s2', 'C2-s2', 'C3-s2']"]}],"metadata":{"interpreter":{"hash":"5ed6376b18e7634873c6ac370ba99f10b16df77a55766a7c88588edd86bf614a"},"kernelspec":{"display_name":"Python 3.6.12 64-bit ('py36': conda)","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.6.13"},"orig_nbformat":4},"nbformat":4,"nbformat_minor":2}
